File: HOT100/6.矩阵/54.螺旋矩阵.cpp
/*
 * @lc app=leetcode.cn id=54 lang=cpp
 *
 * [54] 螺旋矩阵
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        int l = 0, r = n - 1, t = 0, b = m - 1;
        vector<int> res;
        while (1)
        {
            for (int i = l; i <= r; i ++ ) res.push_back(matrix[t][i]);
            if (++ t > b) break;
            for (int i = t; i <= b; i ++ ) res.push_back(matrix[i][r]);
            if (-- r < l) break;
            for (int i = r; i >= l; i -- ) res.push_back(matrix[b][i]);
            if (-- b < t) break;
            for (int i = b; i >= t; i -- ) res.push_back(matrix[i][l]);
            if (++ l > r) break;
        }
        return res;
    }
};
// @lc code=end

File: CodeTop/字节/54.螺旋矩阵.py
#
# @lc app=leetcode.cn id=54 lang=python3
#
# [54] 螺旋矩阵
#
from mytools import *
# @lc code=start
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        res = []
        l, r, t, b = 0, n - 1, 0, m - 1
        while True:
            if t <= b:
                for i in range(l, r + 1):
                    res.append(matrix[t][i])
            else: break
            t += 1
            if r >= l:
                for i in range(t, b + 1):
                    res.append(matrix[i][r])
            else: break
            r -= 1
            if b >= t:
                for i in range(r, l - 1, -1):
                    res.append(matrix[b][i])
            else: break
            b -= 1
            if l <= r:
                for i in range(b, t - 1, -1):
                    res.append(matrix[i][l])
            else: break
            l += 1
        return res
# @lc code=end

